network/retrieval/retrieve.go

```diff
@@ -409,6 +420,7 @@ FINDPEER:
 	}
 
 	ret := &RetrieveRequest{
+		Ruid: uint(rand.Uint32()),
```

Comment:
I want to be absolutely sure that this random number cannot be guessed by peers. The result of peers being able to guess this number is to have chunk deliveries to appear as solicited, while they are not, draining all funds in the chequebook.

The go standard documentation advices to use `crypto/rand` for security-sensitive usages of the random number, which I think is the case right now.

Reply:
Is the communication between peers encrypted?

Reply:
This is an irrelevant assumption to the scope of this PR.
Here's why:
1. Even if you sniff the `ruid` you'd still have to impersonate the node and be connected to the peer in order to hit the same `ruid` in the codebase (since the check is per _peer_)
2. Even if you do the above, you'd still need to provide the actual chunk with the correct hash (this will not fly since check address is matched to the hash generated from the actual chunk content when we get the chunk)
3. `Ruid` is between peers so even if you are forwarding a chunk that another node requested from you, you'll generate a new `Ruid` for the forwarded request

> The go standard documentation advices to use crypto/rand for security-sensitive usages of the random number, which I think is the case right now.

I think that this is mainly geared at very secure operations such as getting a random seed for private key generation, which is certainly not the case here.
All in all I think that `rand` is enough here although I don't really have an opinion whether a change to `crypto/rand` is necessary or not. @janos @nonsense opinions?

Reply:
I would assume that by monitoring messages between peers, it is possible to predict the seed used for ruid by math/rand. This assumption needs to be verified, of course. Using crypto/rand would make that much harder or impossible and I do not think that it would influence the performance in this case compared to math/rand.

If we would ensure encrypted connections, that would not matter, but I do not think that we are encrypting our payload. In this case, we would just need to sign the ruid from one peer to another, and to reject all messages with invalid signatures. This could be an improvement for the future. But points by @acud are correct and we probably do not need this at all.

Reply:
@Eknir just an additional thought: even if the `ruid` and address were spoofed, the balance would be drained towards the spoofed node. it's _still_ an attack, but a less effective one.

Reply:
Thanks for the answers guys. It convinced me everything is good. I think the fact that ruid's are generated per peer is strong enough. Let's say that a node can impersonate other nodes and guess the ruid's and trick the receiving node into writing a cheque to him: even if this is possible, the authentic node will still attempt to send the actual content, causing a disconnect. This essentially sets an upper bound in profit described above and that makes me confident that even while it might be possible (debatable), people will not attempt to do this, as the profit is limited to the number of peers of a certain node times the price for one chunk delivery (very little).

Reply:
We are using RLPx for transport, which is encrypted. I've heard that the protocol is flawed, but I don't know in detail what that means. https://github.com/ethereum/devp2p/blob/master/rlpx.md